### PR TITLE
feat: search bar

### LINKED ISSUE #2286: FE: [Query Builder] Search Bar (Where Clause)
## Is your feature request related to a problem?

*Please describe.*
Need to build a common search bar in which users can do and generate query

https://www.figma.com/file/i2CsZFl27dhmER6gnx08wj/Integ.-QueryBuilder%2B3SigNalCorrelation?node-id=1%3A8036&t=tkzNwNk8Su1ivj1Y-0

fetching of one query tag involves 3 part 

1. fetching of keys and selection of value
1. selection of the operator
1. selection of the value

Points to look for the development

1. for an operator like existing value can be empty
1. while pressing backspace prev tag can be editable and populate options accordingly 
1. key, operator cannot contain space in it
1. User should freely type and update the options accordingly 

Note: when key, value and operator are selected this will turn into an antd Tag https://ant.design/components/tag

### PR DESCRIPTION
**Pull request by KolesnikAntony**
close #2286

-Created context wrapper with form state. It gives us posibility to add antother fields for "query bulder" inside context form.
-Created "search form" component
-Connected API for searching keys and values 
-Added operators for each key
-Added multiply selection as tags of queries
-Added possibilities to choose the tag without of value when we have the operator 'exist' or 'nor exist'
-Added unit tests

For testig of component we can add context wrapper and component as child there you needed;
image

### CODE REVIEW

File: frontend/src/api/queryBuilder/getKeysAutoComplete.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+import { ErrorResponse, SuccessResponse } from 'types/api';
+import { PayloadProps } from 'types/api/queryBuilder/getKeysAutoComplete';
+
+const getKeysAutoComplete = async (
+	searchText: string,
+): Promise<SuccessResponse<PayloadProps[]> | ErrorResponse> => {
+	try {
+		const data = await axios({
+			method: 'get',
+			url: `/api/v3/autocomplete/attribute_keys?aggregateOperator=sum&dataSource=metrics&aggregateAttribute=signoz_calls_total&searchText=${searchText}`,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getKeysAutoComplete.ts:13]: take metrics as parameter

File: frontend/src/api/queryBuilder/getKeysAutoComplete.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+			baseURL: 'http://34.229.125.174:3301',
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getKeysAutoComplete.ts:14]: remove this one

Comment (palashgdev) [reply frontend/src/api/queryBuilder/getKeysAutoComplete.ts:14]: please replace the env

File: frontend/src/api/queryBuilder/getValuesAutoComplete.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+const getValuesAutoComplete = async (
+	attributeKey: string,
+	searchText: string,
+): Promise<SuccessResponse<PayloadProps[]> | ErrorResponse> => {
+	try {
+		const data = await axios({
+			method: 'get',
+			url: `api/v3/autocomplete/attribute_values?aggregateOperator=sum&dataSource=metrics&aggregateAttribute=signoz_calls_total&attributeKey=${attributeKey}&searchText=${searchText}`,
+			baseURL: 'http://34.229.125.174:3301',
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getValuesAutoComplete.ts:15]: same here

File: frontend/src/api/queryBuilder/getValuesAutoComplete.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+import { PayloadProps } from 'types/api/queryBuilder/getKeysAutoComplete';
+
+const getValuesAutoComplete = async (
+	attributeKey: string,
+	searchText: string,
+): Promise<SuccessResponse<PayloadProps[]> | ErrorResponse> => {
+	try {
+		const data = await axios({
+			method: 'get',
+			url: `api/v3/autocomplete/attribute_values?aggregateOperator=sum&dataSource=metrics&aggregateAttribute=signoz_calls_total&attributeKey=${attributeKey}&searchText=${searchText}`,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getValuesAutoComplete.ts:14]: are these parameter constant?

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+	handleFetchOption: (value: string) => void;
+	handleKeyDown: (e: React.KeyboardEvent) => void;
+	options: OptionType[];
+	tags: string[];
+	searchValue: string;
+};
+
+export const useAutoComplete = (): ReturnT => {
+	const { pathname } = useLocation();
+	const operators = useMemo(() => {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:31]: this logic is wrong

Comment (palashgdev) [reply frontend/src/hooks/queryBuilder/useAutoComplete.ts:31]: universal is common to all

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+				setSearchValue('');
+			} else {
+				setSearchValue(value);
+			}
+		}
+	};
+
+	// SET OPTIONS
+	useEffect(() => {
+		if (searchValue) {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:68]: please move this logic to utils

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+	}, [key, keys, operator, operators, results, searchValue]);
+
+	// HANDLE INPUT SEARCH
+	const handleSearch = useCallback((value: string) => {
+		setSearchValue(value);
+	}, []);
+
+	// FETCH OPTIONS
+	const handleFetchOption = useCallback(
+		async (value: string) => {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:98]: this also we are breaking SRP here

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+	// FOUND VALUES
+	const [results, setResults] = useState([]);
+	// OPTIONS
+	const [options, setOptions] = useState<OptionType[]>([]);
+	// SELECTED OPTIONS
+	const [tags, setTags] = useState<string[]>([]);
+
+	const [key, operator] = useSeparateSearchString(searchValue, keys, operators);
+
+	const handleAddTags = (value: string): void => {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:55]: same for this one also

File: frontend/src/container/QueryBuilder/index.tsx [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+const initialState: InitialStateI = {
+	search: '',
+};
+
+const defaultContextValue: ContextValueI = {
+	values: initialState,
+	handleChangeValue: (): HandleChangeReturnT => (): void => {},
+	handleSubmit: () => {},
```
Comment (palashgdev) [inline frontend/src/container/QueryBuilder/index.tsx:33]: VoidFunction

File: frontend/src/container/QueryBuilder/index.tsx [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+const QueryBuilderContext = createContext<ContextValueI>(defaultContextValue);
+const useQueryBuilderContext = (): ContextValueI =>
+	useContext(QueryBuilderContext);
+
+function QueryBuilderContainer({
+	children,
+}: QueryBuilderContainerPropsI): JSX.Element {
+	const [values, setValues] = useState<InitialStateI>(initialState);
+
+	const handleChangeValue = useCallback(
```
Comment (palashgdev) [inline frontend/src/container/QueryBuilder/index.tsx:45]: onChangeHandler

File: frontend/src/container/QueryBuilder/index.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+
+	const handleSubmit = useCallback(() => {
+		console.log(values, '----values');
+	}, [values]);
+
+	const providerValue = useMemo(
+		() => ({
+			values,
+			handleChangeValue,
+			handleSubmit,
```
Comment (palashgdev) [inline frontend/src/container/QueryBuilder/index.tsx:60]: onSubmitHandler

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+	NOT_EXISTS,
+	QUERY_BUILDER_OPERATORS,
+} from 'constants/queryBuilder';
+import ROUTES from 'constants/routes';
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+import { useLocation } from 'react-router-dom';
+
+import { useSeparateSearchString } from './useSeparateSearchString';
+
+type OptionType = {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:14]: Option

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+	handleKeyDown: (e: React.KeyboardEvent) => void;
+	options: OptionType[];
+	tags: string[];
+	searchValue: string;
+};
+
+export const useAutoComplete = (): ReturnT => {
+	const { pathname } = useLocation();
+	const operators = useMemo(() => {
+		switch (pathname) {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:32]: donot switch case use mapper function for same

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/api/queryBuilder/getValuesAutoComplete.ts [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+	searchText: string,
+	aggregateOperator: undefined | string = 'sum',
+	dataSource: undefined | string = 'metrics',
+	aggregateAttribute: undefined | string = 'signoz_calls_total',
+): Promise<SuccessResponse<PayloadProps> | ErrorResponse> => {
+	try {
+		const data = await axios({
+			method: 'get',
+			url: `api/v3/autocomplete/attribute_values?aggregateOperator=${aggregateOperator}&dataSource=${dataSource}&aggregateAttribute=${aggregateAttribute}&attributeKey=${attributeKey}&searchText=${searchText}`,
+			baseURL: process.env.QUERY_BUILDER_ENDPOINT,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getValuesAutoComplete.ts:18]: let's use `FRONTEND_API_ENDPOINT` so we can remove this baseUrl

File: frontend/src/api/queryBuilder/getKeysAutoComplete.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+	dataSource: undefined | string = 'metrics',
+	aggregateAttribute: undefined | string = 'signoz_calls_total',
+): Promise<SuccessResponse<PayloadProps[]> | ErrorResponse> => {
+	try {
+		const data = await axios({
+			method: 'get',
+			url: `/api/v3/autocomplete/attribute_keys?aggregateOperator=${aggregateOperator}&dataSource=${dataSource}&aggregateAttribute=${aggregateAttribute}&searchText=${
+				searchText || ''
+			}`,
+			baseURL: process.env.QUERY_BUILDER_ENDPOINT,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getKeysAutoComplete.ts:19]: same here also

File: frontend/src/components/QueryBuilder/Search/index.tsx [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+import { QUERY_BUILDER_STATE_KEYS } from 'constants/queryBuilder';
+import { useQueryBuilderContext } from 'container/QueryBuilder';
+import { useAutoComplete } from 'hooks/queryBuilder/useAutoComplete';
+import React, { useEffect } from 'react';
+
+import { Tag } from './tag';
+
+function QueryBuilderSearch(): JSX.Element {
+	const { onChangeHandler, onSubmitHandler } = useQueryBuilderContext();
+	const { Option } = Select;
```
Comment (palashgdev) [inline frontend/src/components/QueryBuilder/Search/index.tsx:12]: donot define here. as rerendering would be re-calculating this object... so use Select.Option directly

File: frontend/src/components/QueryBuilder/Search/index.tsx [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+				onSearch={handleSearch}
+				searchValue={searchValue}
+				style={{ width: '100%' }}
+			>
+				{options.map((o) => (
+					<Option
+						key={o.value}
+						value={o.value}
+						className={
+							o.selected
```
Comment (palashgdev) [inline frontend/src/components/QueryBuilder/Search/index.tsx:55]: donot write inline logic

File: frontend/src/components/QueryBuilder/Search/index.tsx [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+				autoClearSearchValue={false}
+				mode="multiple"
+				placeholder="Search Filter"
+				value={tags}
+				onDeselect={handleClearTag}
+				onSelect={handleSelect}
+				onInputKeyDown={handleKeyDown}
+				onSearch={handleSearch}
+				searchValue={searchValue}
+				style={{ width: '100%' }}
```
Comment (palashgdev) [inline frontend/src/components/QueryBuilder/Search/index.tsx:48]: make styled Select which has `100%` width

File: frontend/src/components/QueryBuilder/Search/tag.tsx [hunk 19]
```diff
@@ -0,0 +1,9 @@
+import { Tag as AntTag } from 'antd';
+import styled from 'styled-components';
+
+export const Tag = styled(AntTag)`
+	font-style: normal;
+	font-weight: 300;
+	font-size: 12px;
```
Comment (palashgdev) [inline frontend/src/components/QueryBuilder/Search/tag.tsx:7]: rem

File: frontend/src/container/QueryBuilder/index.tsx [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+	children: React.ReactElement | React.ReactNode;
+}
+
+interface InitialStateI {
+	search: string;
+}
+
+type VoidFunction = (value: string) => void;
+
+interface ContextValueI {
```
Comment (palashgdev) [inline frontend/src/container/QueryBuilder/index.tsx:20]: create type.ts and put all types there

File: frontend/src/container/QueryBuilder/index.tsx [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+	values: InitialStateI;
+	onChangeHandler: (type: QueryBuilderStateT) => VoidFunction;
+	onSubmitHandler: () => void;
```
Comment (palashgdev) [inline frontend/src/container/QueryBuilder/index.tsx:23]:
```suggestion
	onSubmitHandler: VoidFunction;
```

File: frontend/src/container/QueryBuilder/index.tsx [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+const initialState: InitialStateI = {
+	search: '',
+};
+
+const defaultContextValue: ContextValueI = {
+	values: initialState,
+	onChangeHandler: (): VoidFunction => (): void => {},
+	onSubmitHandler: (): void => {},
```
Comment (palashgdev) [inline frontend/src/container/QueryBuilder/index.tsx:33]:
```suggestion
	onSubmitHandler: VoidFunction,
```

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 23]
```diff
@@ -0,0 +1,174 @@
...
+	const { handleAddTag, handleClearTag, tags } = useTag(
+		key,
+		isValidTag,
+		isFreeText,
+		handleSearch,
+	);
+
+	// GET OPERATORS BY KEY TYPE
+	const operators = useMemo(() => {
+		const currentKey = keys.find((el) => el.key === key);
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:67]: let's make a util for this one

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+		getKeysAutoComplete().then(({ payload }) => {
+			if (payload) {
+				setKeys(payload);
+			}
+		});
+	}, []);
+
+	const handleSetKey = (payload: PayloadProps[] | null): void => {
+		if (payload) {
+			setKeys(payload as []);
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:31]: do we need as `[]`?

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts [hunk 25]
```diff
@@ -0,0 +1,81 @@
...
+					}
+				}
+			}
+		}
+	};
+
+	const clearFetcher = useRef(handleFetchOption).current;
+
+	useEffect(() => {
+		const timer = setTimeout(() => clearFetcher(searchValue).then(), 100);
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:71]: can you please check if there's already a hook for the same in react-use

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 26]
```diff
@@ -0,0 +1,174 @@
...
+		setSearchValue(value);
+	}, []);
+
+	const [options, setOptions] = useState<Option[]>([]);
+
+	// GET SUGGESTION KEYS AND VALUES
+	const { keys, results } = useFetchKeysAndValues(searchValue);
+
+	// SELECT KEY, OPERATOR AND RESULT
+	const [key, operator, result] = useSetCurrentKeyAndOperator(searchValue, keys);
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:46]: we can write jsdoc description to be more instead of inline comments

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 27]
```diff
@@ -0,0 +1,174 @@
...
+};
+
+export type KeyType = {
+	key: string;
+	dataType: 'string' | 'boolean' | 'number';
+	type: string;
+};
+
+export const useAutoComplete = (): ReturnT => {
+	const [searchValue, setSearchValue] = useState('');
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:33]: can you check if we can use antd form for this one

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/container/QueryBuilder/index.tsx (hunk 22 above)
Comment (KolesnikAntony) [reply frontend/src/container/QueryBuilder/index.tsx:33]: it is not type or interface . This is initial value. We cannot use type as function here.

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts (hunk 25 above)
Comment (KolesnikAntony) [reply frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:71]: There is the similar. But it doesn't work for us

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts (hunk 27 above)
Comment (KolesnikAntony) [reply frontend/src/hooks/queryBuilder/useAutoComplete.ts:33]: We can but I think this is not better idea on this level. We also have general context with values state and there we can use it but I can see why. But if you want, of course I can rewrite it on ant form

File: frontend/src/components/QueryBuilder/Search/tag.tsx (hunk 19 above)
Comment (KolesnikAntony) [reply frontend/src/components/QueryBuilder/Search/tag.tsx:7]: it is with rem 
image

File: frontend/README.md [hunk 28]
```diff
@@ -12,7 +12,7 @@
 or
 `docker build . -t tagname`
 
-**Tag to remote url- Introduce versinoing later on**
+**Styles to remote url- Introduce versinoing later on**
```
Comment (palashgdev) [inline frontend/README.md:15]: please remove this change

File: frontend/src/api/queryBuilder/getKeysAutoComplete.ts [hunk 29]
```diff
@@ -0,0 +1,33 @@
...
+	dataSource: undefined | string = 'metrics',
+	aggregateAttribute: undefined | string = 'signoz_calls_total',
+): Promise<SuccessResponse<PayloadProps[]> | ErrorResponse> => {
+	try {
+		const data = await axios({
+			method: 'get',
+			url: `/api/v3/autocomplete/attribute_keys?aggregateOperator=${aggregateOperator}&dataSource=${dataSource}&aggregateAttribute=${aggregateAttribute}&searchText=${
+				searchText || ''
+			}`,
+			baseURL: process.env.FRONTEND_API_ENDPOINT,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getKeysAutoComplete.ts:19]: baseUrl is already configure in the Axios Instance no need to configure this one

File: frontend/src/api/queryBuilder/getValuesAutoComplete.ts [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+	attributeKey: string,
+	searchText: string,
+	aggregateOperator: undefined | string = 'sum',
+	dataSource: undefined | string = 'metrics',
+	aggregateAttribute: undefined | string = 'signoz_calls_total',
+): Promise<SuccessResponse<PayloadProps> | ErrorResponse> => {
+	try {
+		const data = await axios({
+			method: 'get',
+			url: `api/v3/autocomplete/attribute_values?aggregateOperator=${aggregateOperator}&dataSource=${dataSource}&aggregateAttribute=${aggregateAttribute}&attributeKey=${attributeKey}&searchText=${searchText}`,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getValuesAutoComplete.ts:17]: let's use the URL class instance to add the params

File: frontend/src/api/queryBuilder/getValuesAutoComplete.ts [hunk 31]
```diff
@@ -0,0 +1,32 @@
...
+			baseURL: process.env.FRONTEND_API_ENDPOINT,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getValuesAutoComplete.ts:18]: same here also

File: frontend/src/container/QueryBuilder/index.tsx (hunk 22 above)
Comment (palashgdev) [reply frontend/src/container/QueryBuilder/index.tsx:25]: make sense

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts (hunk 25 above)
Comment (palashgdev) [reply frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:69]: why ?

File: frontend/src/hooks/queryBuilder/useTag.ts [hunk 32]
```diff
@@ -0,0 +1,33 @@
+import { useCallback, useState } from 'react';
+
+type ReturnT = {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useTag.ts:3]: we can add better name for this one 😅

File: frontend/src/hooks/queryBuilder/useTagValidation.ts [hunk 33]
```diff
@@ -0,0 +1,67 @@
...
+	isMulti: boolean;
+	isFreeText: boolean;
+};
+
+export const useTagValidation = (
+	value: string,
+	operator: string,
+	result: string[],
+): ReturnT => {
+	const operatorType = useMemo(() => {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useTagValidation.ts:19]: move this to utils

File: frontend/src/hooks/queryBuilder/useTagValidation.ts [hunk 34]
```diff
@@ -0,0 +1,67 @@
...
+			case OPERATORS.NOT_CONTAINS:
+			case OPERATORS.EQUALS:
+			case OPERATORS.NOT_EQUALS:
+				return QUERY_BUILDER_SEARCH_VALUES.SINGLE;
+			default:
+				return QUERY_BUILDER_SEARCH_VALUES.NOT_VALID;
+		}
+	}, [operator]);
+
+	const isValidTag = useMemo(() => {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useTagValidation.ts:43]: please make a utils for this one

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/api/queryBuilder/getValuesAutoComplete.ts [hunk 35]
```diff
@@ -0,0 +1,30 @@
...
+const getValuesAutoComplete = async (
+	attributeKey: string,
+	searchText: string,
+	aggregateOperator: undefined | string = 'sum',
+	dataSource: undefined | string = 'metrics',
+	aggregateAttribute: undefined | string = 'signoz_calls_total',
+): Promise<SuccessResponse<PayloadProps> | ErrorResponse> => {
+	try {
+		const data = await ApiV3Instance.get(
+			`/autocomplete/attribute_values?aggregateOperator=${aggregateOperator}&dataSource=${dataSource}&aggregateAttribute=${aggregateAttribute}&attributeKey=${attributeKey}&searchText=${searchText}`,
```
Comment (palashgdev) [inline frontend/src/api/queryBuilder/getValuesAutoComplete.ts:16]: this is not scalable 

let's make a utils do this job
 
https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams/set#return_value

File: frontend/src/components/QueryBuilder/Search/index.tsx [hunk 36]
```diff
@@ -0,0 +1,69 @@
...
+				mode="multiple"
+				placeholder="Search Filter"
+				value={tags}
+				onDeselect={handleClearTag}
+				onSelect={handleSelect}
+				onInputKeyDown={handleKeyDown}
+				onSearch={handleSearch}
+				searchValue={searchValue}
+			>
+				{options.map((o) => (
```
Comment (palashgdev) [inline frontend/src/components/QueryBuilder/Search/index.tsx:54]:
```suggestion
				{options.map((option) => (
```

File: frontend/src/components/QueryBuilder/Search/styles.tsx [hunk 37]
```diff
@@ -0,0 +1,15 @@
+import { Select as AtnSelect, Tag as AntTag } from 'antd';
+import styled from 'styled-components';
+
+const Tag = styled(AntTag)`
+	font-style: normal;
+	font-weight: 300;
+	font-size: 12px;
```
Comment (palashgdev) [inline frontend/src/components/QueryBuilder/Search/styles.tsx:7]: 0.75 rem

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts (hunk 6 above)
Comment (palashgdev) [reply frontend/src/hooks/queryBuilder/useAutoComplete.ts:105]: which will give the options and setting here

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 38]
```diff
@@ -0,0 +1,212 @@
...
+		results,
+		searchValue,
+	]);
+
+	/**
+	 * Handle select of options
+	 */
+	const handleSelect = (value: string): void => {
+		if (isMulti) {
+			setSearchValue((prev) => {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:145]: same for this one please move the logic to utils which will give searchValue

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 39]
```diff
@@ -0,0 +1,212 @@
...
+
+		if (e.key === 'Backspace' && !searchValue) {
+			e.stopPropagation();
+			const last = tags[tags.length - 1];
+			setSearchValue(last);
+			handleClearTag(last);
+		}
+	};
+
+	const isFilter = useMemo(() => !isMulti, [isMulti]);
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:189]: no need to use Memo on primitive values

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 40]
```diff
@@ -0,0 +1,212 @@
...
+
+	const optionsUpdated = useMemo(
+		() =>
+			options.map((o) => {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:193]:
```suggestion
			options.map((option) => {
```

File: frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts [hunk 41]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+export const useFetchKeysAndValues = (
+	searchValue: string,
+): UseFetchKeysAndValuesReturnT => {
+	const [keys, setKeys] = useState<KeyType[]>([]);
+	const [results, setResults] = useState([]);
+
+	useEffect(() => {
+		getKeysAutoComplete().then(({ payload }) => {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useFetchKeysAndValues.ts:24]: let's use useQuery here for better handling

**Review by palashgdev — CHANGES_REQUESTED** [review]

File: frontend/src/hooks/queryBuilder/useAutoComplete.ts [hunk 42]
```diff
@@ -0,0 +1,209 @@
...
+		handleSearch,
+	);
+
+	const operators = useOperators(key, keys);
+
+	/**
+	 * Set options based on the parameters
+	 */
+	useEffect(() => {
+		if (!key) {
```
Comment (palashgdev) [inline frontend/src/hooks/queryBuilder/useAutoComplete.ts:105]: please move this logic to utils

**Review by palashgdev — CHANGES_REQUESTED** [review]

**Pull request closed without merging**